### PR TITLE
sql: allow SNAPSHOT isolation and add version gate for isolation levels

### PR DESCRIPTION
**Pull request by rafiss**
### sql: allow SNAPSHOT isolation if hidden setting is enabled

If the `sql.txn.snapshot_isolation.enabled` cluster setting is set to true, then SNAPSHOT
isolation can now be used. This is intended for usage only by CockroachDB developers, so
it is not exposed publicly.

---

### sql: add version gate for customizing isolation level

As a defensive measure, the transaction isolation level should only be
configurable if the 23.2 upgrade is finalized. This lets us avoid having
to reason about READ COMMITTED or SNAPSHOT transactions for clusters
where some nodes are still on v23.1.

---

###  sql: remove dead code for showing transaction_isolation

There's no need to check for UnspecifiedIsolation, since the kv txn
always has its isolation level defined explicitly.

---

###  parser: map READ UNCOMMITTED and REPEATABLE READ isolation levels

These get mapped to the next closest isolation level that is actually
supported. By default, REPEATABLE READ is mapped to SERIALIZABLE, but if the
hidden setting that enables SNAPSHOT is on, then REPEATABLE READ is
mapped to SNAPSHOT.

Epic: CRDB-26545
Release note: None

### CODE REVIEW

**Review by nvb — COMMENTED** [review]
> Reviewed 6 of 6 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103488)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rafiss)

___
*[`pkg/sql/conn_executor.go` line 3333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NVkzC5d78MrZwpEw7BL:-NVkzC5d78MrZwpEw7BM:b-dergfj) ([raw file](https://github.com/cockroachdb/cockroach/blob/69240c0e685d91d566d2bfea05cfe66ecee53670/pkg/sql/conn_executor.go#L3333)):*
> ```Go
> 
> func (ex *connExecutor) txnIsolationLevelToKV(level tree.IsolationLevel) isolation.Level {
> 	allowSnapshot := allowSnapshotIsolation.Get(&ex.server.cfg.Settings.SV)
> ```

nit: consider pushing this cluster setting read into the `case tree.SnapshotIsolation`. An atomic read is cheap, but not free.
___
*[`pkg/sql/set_session_characteristics.go` line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NVkzKxZD3MZ1-BlIwe9:-NVkzKxZD3MZ1-BlIweA:b-3v0zy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/69240c0e685d91d566d2bfea05cfe66ecee53670/pkg/sql/set_session_characteristics.go#L33)):*
> ```Go
> 		case tree.SnapshotIsolation:
> 			level := tree.SerializableIsolation
> 			if allowSnapshotIsolation.Get(&p.execCfg.Settings.SV) {
> ```

Same question about whether this is needed.
___
*[`pkg/sql/vars.go` line 415 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NVkz20D0PAmpaqEWcD6:-NVkz20D0PAmpaqEWcD7:b-x8w8zp) ([raw file](https://github.com/cockroachdb/cockroach/blob/69240c0e685d91d566d2bfea05cfe66ecee53670/pkg/sql/vars.go#L415)):*
> ```Go
> 				}
> 			}
> 			if level == tree.SnapshotIsolation && !allowSnapshotIsolation.Get(&m.settings.SV) {
> ```

Do we need this? Won't the default isolation level make its way through `txnIsolationLevelToKV`?
___
*[`pkg/sql/parser/sql.y` line 6889 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NVky7OjAR42HGRTZwzJ:-NVky7OjAR42HGRTZwzK:b-gviaf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/69240c0e685d91d566d2bfea05cfe66ecee53670/pkg/sql/parser/sql.y#L6889)):*
> ```Yacc
>   READ UNCOMMITTED
>   {
>     $$.val = tree.SerializableIsolation
> ```

Should we go ahead and add `ReadUncommitted` and `RepeatableRead` to the enum so that we can perform this mapping up in `txnIsolationLevelToKV` instead of here? I'm asking both because we'll want to map `ReadUncommitted` to `ReadCommitted` and also because we'll want to map `RepeatableRead` to `Snapshot` (which may or may not be mapped to `Serializable`, depending on the new setting).
___
*[`pkg/sql/sem/tree/txn.go` line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NVkyfYNDvY9afvSFPjP:-NVkyfYNDvY9afvSFPjQ:bgec844) ([raw file](https://github.com/cockroachdb/cockroach/blob/69240c0e685d91d566d2bfea05cfe66ecee53670/pkg/sql/sem/tree/txn.go#L27)):*
> ```Go
> const (
> 	UnspecifiedIsolation IsolationLevel = iota
> 	SerializableIsolation
> ```

If the values in this enum don't matter, let's put them in increasing strength order (the order we have in the grammar).

**Review by michae2 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103488)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @rafiss)

___
*[`pkg/sql/conn_executor.go` line 3333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NVkzC5d78MrZwpEw7BL:-NVl1k3I3TaIssru-Ogn:bru1es9) ([raw file](https://github.com/cockroachdb/cockroach/blob/69240c0e685d91d566d2bfea05cfe66ecee53670/pkg/sql/conn_executor.go#L3333)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: consider pushing this cluster setting read into the `case tree.SnapshotIsolation`. An atomic read is cheap, but not free.
</blockquote></details>

(Or maybe consider lifting it up into the callers of this function, guarded by a conditional, so that this can remain a pure function.)
___
*[`pkg/sql/parser/sql.y` line 6889 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NVky7OjAR42HGRTZwzJ:-NVl2ipm2BMZkTjSeaHF:bps2l1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/69240c0e685d91d566d2bfea05cfe66ecee53670/pkg/sql/parser/sql.y#L6889)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we go ahead and add `ReadUncommitted` and `RepeatableRead` to the enum so that we can perform this mapping up in `txnIsolationLevelToKV` instead of here? I'm asking both because we'll want to map `ReadUncommitted` to `ReadCommitted` and also because we'll want to map `RepeatableRead` to `Snapshot` (which may or may not be mapped to `Serializable`, depending on the new setting).
</blockquote></details>

It does seem like the AST should have `ReadUncommitted` so that we can print out the SQL statement as it was entered, even if the mapping changes it to a different isolation level.

**Review by michae2 — APPROVED** [review]
Those were just nits for me, :lgtm:

> Reviewed 6 of 6 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-:-NVmWMo72P4kk6cXUCes:brx8svc)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @rafiss)

**Review by michae2 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103488)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @rafiss)

___
*[`pkg/sql/conn_executor.go` line 3326 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NW3h4VIBKGO0z0ofdV4:-NW3h4VIBKGO0z0ofdV5:btyx5nv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f2a2b234cfe8b22764934a10db8c8e333fc927c/pkg/sql/conn_executor.go#L3326)):*
> ```Go
> var allowSnapshotIsolation = settings.RegisterBoolSetting(
> 	settings.TenantWritable,
> 	"sql.txn.snapshot_isolation.enabled",
> ```

This is just bikeshedding, but would it make more sense for this to be a session setting rather than a cluster setting?

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103488)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/sql/conn_executor.go` line 3333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NVkzC5d78MrZwpEw7BL:-NW9RKaYCAugsn4Awjor:bigqxee) ([raw file](https://github.com/cockroachdb/cockroach/blob/69240c0e685d91d566d2bfea05cfe66ecee53670/pkg/sql/conn_executor.go#L3333)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

(Or maybe consider lifting it up into the callers of this function, guarded by a conditional, so that this can remain a pure function.)
</blockquote></details>

i don't see a huge advantage to making this pure, and there is an advantage to limiting the places we have to check the setting, so i'll go with nathan's suggestion.
___
*[`pkg/sql/conn_executor.go` line 3326 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NW3h4VIBKGO0z0ofdV4:-NW9LxYg-elyn9a0XDnf:b-q2rbhg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f2a2b234cfe8b22764934a10db8c8e333fc927c/pkg/sql/conn_executor.go#L3326)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

This is just bikeshedding, but would it make more sense for this to be a session setting rather than a cluster setting?
</blockquote></details>

i'm fine either way. one difference is that cluster settings are limited to usage by admins or anyone with the MODIFYCLUSTERSETTING privilege, while session variables are always settable by anyone. i guess this since is going to be hidden anyway, maybe this matters less?
___
*[`pkg/sql/set_session_characteristics.go` line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NVkzKxZD3MZ1-BlIwe9:-NW9RxwG8HoZjQaSuf_k:b-1cgm84) ([raw file](https://github.com/cockroachdb/cockroach/blob/69240c0e685d91d566d2bfea05cfe66ecee53670/pkg/sql/set_session_characteristics.go#L33)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Same question about whether this is needed.
</blockquote></details>

this code guards `default_transaction_isolation` - this controls the default isolation level that is used if `BEGIN` does not specify one explicitly. if we did not guard this here then it would be possible to set `default_transaction_isolation` to SNAPSHOT even when the setting is disabled, but the changes i've made to `txnIsolationLevelToKV` would cause new transactions to end up using SERIALIZABLE

so basically, this isn't needed for preventing transactions from using SNAPSHOT, but it does prevent some potential UX confusion.
___
*[`pkg/sql/vars.go` line 415 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NVkz20D0PAmpaqEWcD6:-NW9Th3y6kVb5ZufD0R3:bx3pr4w) ([raw file](https://github.com/cockroachdb/cockroach/blob/69240c0e685d91d566d2bfea05cfe66ecee53670/pkg/sql/vars.go#L415)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need this? Won't the default isolation level make its way through `txnIsolationLevelToKV`?
</blockquote></details>

same as above (annoyingly PG has two ways of configuring the default isolation level: the session variable and `SET SESSION CHARACTERISTICS`)
___
*[`pkg/sql/parser/sql.y` line 6889 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NVky7OjAR42HGRTZwzJ:-NW9Tyun0lH3R1W94P1s:b9njbky) ([raw file](https://github.com/cockroachdb/cockroach/blob/69240c0e685d91d566d2bfea05cfe66ecee53670/pkg/sql/parser/sql.y#L6889)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

It does seem like the AST should have `ReadUncommitted` so that we can print out the SQL statement as it was entered, even if the mapping changes it to a different isolation level.
</blockquote></details>

that makes sense. i'll do this
___
*[`pkg/sql/sem/tree/txn.go` line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NVkyfYNDvY9afvSFPjP:-NW9U08YAnhaKc2IPPEy:bb7ohn6) ([raw file](https://github.com/cockroachdb/cockroach/blob/69240c0e685d91d566d2bfea05cfe66ecee53670/pkg/sql/sem/tree/txn.go#L27)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If the values in this enum don't matter, let's put them in increasing strength order (the order we have in the grammar).
</blockquote></details>

sgtm

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103488)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/sql/sem/tree/txn.go` line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NVkyfYNDvY9afvSFPjP:-NWDTxao2LTZqnh3wItI:b-wp4g8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/69240c0e685d91d566d2bfea05cfe66ecee53670/pkg/sql/sem/tree/txn.go#L27)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

sgtm
</blockquote></details>

@nvanbenschoten one thing i wanted to check - the grammar file indicates that `REPEATABLE READ` is higher than `SNAPSHOT`. so then why do we want to have `REPEATABLE READ` get mapped to `SNAPSHOT`? is it because in postgres, the two are actually equivalent? (it seems that each DBMS defines them slightly differently or not at all)

**Review by nvb — COMMENTED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103488)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @rafiss)

___
*[`pkg/sql/sem/tree/txn.go` line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NVkyfYNDvY9afvSFPjP:-NWDV0aP9-mKM8OV3Gq0:b-lec0aa) ([raw file](https://github.com/cockroachdb/cockroach/blob/69240c0e685d91d566d2bfea05cfe66ecee53670/pkg/sql/sem/tree/txn.go#L27)):*
Repeatable Read and Snapshot aren't stronger or weaker than one another. From the RC RFC:
> `REPEATABLE READ` is the next strongest standard isolation level. Formally, the isolation level (PL-2.99 in Adya) permits Phantom Reads but does not permit Write Skew. In its place and under the same name, PostgreSQL implements Snapshot isolation (PL-SI in Adya). Snapshot isolation does not permit Phantom Reads but does permit Write Skew. This deviation is not strictly wrong from a standards conformance perspective due to ANSI SQL's ambiguous definitions (which is what prompted Berenson, Bernstein et al.'s [critique](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/tr-95-51.pdf)), but it is confusing.

Also, see https://jepsen.io/consistency.

So there's no real ordering between the two in terms of strength. In practice, the two are so similar that systems that have an easier time implementing Snapshot than Repeatable Read (namely, MVCC systems) will implement Snapshot and then call in Repeatable Read so that they can claim to be ANSI-SQL compliant.

**Review by nvb — APPROVED** [review]
> Reviewed 2 of 5 files at r4, 3 of 3 files at r5, 1 of 1 files at r6, 8 of 8 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103488)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2 and @rafiss)

___
*[`pkg/sql/conn_executor.go` line 3344 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NWEvfuzDjYkOH56BeFo:-NWEvfuzDjYkOH56BeFp:b-o1v0vj) ([raw file](https://github.com/cockroachdb/cockroach/blob/9be247caa283ade5018c3170c6e1e68226ace1f4/pkg/sql/conn_executor.go#L3344)):*
> ```Go
> 		switch level {
> 		case tree.ReadUncommittedIsolation:
> 			fallthrough
> ```

Consider adding comments in these `fallthrough` cases to reinforce to a reader that what we're doing (mapping one isolation level to another) is intentional. You could also point to https://www.postgresql.org/docs/current/transaction-iso.html to explain that we are doing the same thing as PG.
___
*[`pkg/sql/conn_executor.go` line 3356 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NWEw4dmEreZFl9uRagU:-NWEw4dmEreZFl9uRagV:b96pd0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/9be247caa283ade5018c3170c6e1e68226ace1f4/pkg/sql/conn_executor.go#L3356)):*
> ```Go
> 				ret = isolation.Serializable
> 			}
> 		case tree.SerializableIsolation:
> ```

Thanks for keeping the cases ordered :smiley: 
___
*[`pkg/sql/set_session_characteristics.go` line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NVkzKxZD3MZ1-BlIwe9:-NWEv1kz83BuONqyfve6:b-6baeib) ([raw file](https://github.com/cockroachdb/cockroach/blob/69240c0e685d91d566d2bfea05cfe66ecee53670/pkg/sql/set_session_characteristics.go#L33)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this code guards `default_transaction_isolation` - this controls the default isolation level that is used if `BEGIN` does not specify one explicitly. if we did not guard this here then it would be possible to set `default_transaction_isolation` to SNAPSHOT even when the setting is disabled, but the changes i've made to `txnIsolationLevelToKV` would cause new transactions to end up using SERIALIZABLE

so basically, this isn't needed for preventing transactions from using SNAPSHOT, but it does prevent some potential UX confusion.
</blockquote></details>

Got it, thanks for the clear explanation.
___
*[`pkg/sql/vars.go` line 1505 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NWEv5hfCN7jQvHwCMK1:-NWEv5hfCN7jQvHwCMK2:b6i8m5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/32dbcdb0b68b8dc69dface53426a9fff50776956/pkg/sql/vars.go#L1505)):*
> ```Go
> 				} else {
> 					allowedValues = []string{"serializable", "read committed"}
> ```

nit: stray line

**Review by rafiss — COMMENTED** [review]
tftr!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-:-NWGUDrsCdzo5gNKMPAa:bjpalj3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/sql/conn_executor.go` line 3344 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NWEvfuzDjYkOH56BeFo:-NWGSNhm2QFuLrG08rgL:b3g5a32) ([raw file](https://github.com/cockroachdb/cockroach/blob/9be247caa283ade5018c3170c6e1e68226ace1f4/pkg/sql/conn_executor.go#L3344)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider adding comments in these `fallthrough` cases to reinforce to a reader that what we're doing (mapping one isolation level to another) is intentional. You could also point to https://www.postgresql.org/docs/current/transaction-iso.html to explain that we are doing the same thing as PG.
</blockquote></details>

that's a good idea - added.
___
*[`pkg/sql/conn_executor.go` line 3356 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NWEw4dmEreZFl9uRagU:-NWGTbL197y2sC0iNIyd:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9be247caa283ade5018c3170c6e1e68226ace1f4/pkg/sql/conn_executor.go#L3356)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Thanks for keeping the cases ordered :smiley: 
</blockquote></details>

:+1: 
___
*[`pkg/sql/vars.go` line 1505 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/103488#-NWEv5hfCN7jQvHwCMK1:-NWGToXQEUNdAif04sSn:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/32dbcdb0b68b8dc69dface53426a9fff50776956/pkg/sql/vars.go#L1505)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: stray line
</blockquote></details>

fixed
